File: backend/app/schemas/b1f_inspection.py
"""
整棟工務每日巡檢 - B1F  Pydantic Schemas【寬表格 Pivot 架構】
"""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel


class B1FInspectionBatchOut(BaseModel):
    ragic_id:        str
    inspection_date: str
    inspector_name:  str
    start_time:      str
    end_time:        str
    work_hours:      str
    item_count:      int = 0
    synced_at:       Optional[datetime] = None

    class Config:
        from_attributes = True


class B1FInspectionItemOut(BaseModel):
    ragic_id:       str
    batch_ragic_id: str
    seq_no:         int
    item_name:      str
    result_raw:     str
    result_status:  str
    abnormal_flag:  bool
    synced_at:      Optional[datetime] = None

    class Config:
        from_attributes = True


class B1FInspectionBatchKPI(BaseModel):
    total:            int   = 0
    normal:           int   = 0
    abnormal:         int   = 0
    pending:          int   = 0
    unchecked:        int   = 0
    completion_rate:  float = 0.0
    normal_rate:      float = 0.0


class StatusDistItem(BaseModel):
    status: str
    label:  str
    count:  int
    color:  str


class B1FInspectionBatchDetail(BaseModel):
    batch: B1FInspectionBatchOut
    kpi:   B1FInspectionBatchKPI
    items: List[B1FInspectionItemOut]


class B1FInspectionStats(BaseModel):
    latest_batch:        Optional[B1FInspectionBatchOut] = None
    latest_kpi:          Optional[B1FInspectionBatchKPI] = None
    recent_abnormal:     List[B1FInspectionItemOut] = []
    recent_pending:      List[B1FInspectionItemOut] = []
    status_distribution: List[StatusDistItem]       = []
    total_batches_7d:    int = 0
    abnormal_trend:      List[dict] = []

File: frontend/src/types/b1fInspection.ts
/**
 * 整棟工務每日巡檢 - B1F  TypeScript 型別定義【寬表格 Pivot 架構】
 *
 * 對應 Ragic Sheet 4（full-building-inspection/4）
 * 設備欄位由同步服務動態偵測，數量依 Sheet 實際欄位而定
 */

export type B1FInspectionResultStatus = 'normal' | 'abnormal' | 'pending' | 'unchecked' | 'no_record'

export interface B1FInspectionBatch {
  ragic_id:        string
  inspection_date: string
  inspector_name:  string
  start_time:      string
  end_time:        string
  work_hours:      string
  item_count:      number
  synced_at?:      string | null
}

export interface B1FInspectionItem {
  ragic_id:       string
  batch_ragic_id: string
  seq_no:         number
  item_name:      string
  result_raw:     string
  result_status:  B1FInspectionResultStatus
  abnormal_flag:  boolean
  synced_at?:     string | null
}

export interface B1FInspectionBatchKPI {
  total:            number
  normal:           number
  abnormal:         number
  pending:          number
  unchecked:        number
  completion_rate:  number
  normal_rate:      number
}

export interface B1FStatusDistItem {
  status: string
  label:  string
  count:  number
  color:  string
}

export interface B1FInspectionBatchListItem {
  batch: B1FInspectionBatch
  kpi:   B1FInspectionBatchKPI
}

export interface B1FInspectionBatchDetail {
  batch: B1FInspectionBatch
  kpi:   B1FInspectionBatchKPI
  items: B1FInspectionItem[]
}

export interface B1FAbnormalTrendItem {
  date:           string
  abnormal_count: number
  has_record:     boolean
}

export interface B1FInspectionStats {
  latest_batch:        B1FInspectionBatch | null
  latest_kpi:          B1FInspectionBatchKPI | null
  recent_abnormal:     B1FInspectionItem[]
  recent_pending:      B1FInspectionItem[]
  status_distribution: B1FStatusDistItem[]
  total_batches_7d:    number
  abnormal_trend:      B1FAbnormalTrendItem[]
}

export interface B1FInspectionDailySummary {
  inspection_date: string
  inspector_name:  string
  start_time:      string
  result_status:   B1FInspectionResultStatus
  result_raw:      string
  abnormal_flag:   boolean
  has_record:      boolean
  is_today:        boolean
}

export interface B1FInspectionItemHistory {
  item_name:     string
  daily_summary: B1FInspectionDailySummary[]
  stats: {
    total_days:    number
    normal_days:   number
    abnormal_days: number
  }
}
